server/src/main/java/io/crate/execution/engine/indexing/ShardingUpsertExecutor.java

```diff
@@ -252,6 +254,7 @@ private boolean shouldPauseOnPartitionCreation(ShardedRequests<ShardUpsertReques
 
     @Override
     public CompletableFuture<? extends Iterable<Row>> apply(BatchIterator<Row> batchIterator) {
+        var isUsedBytesOverThreshold = new IsUsedBytesOverThreshold(queryCircuitBreaker);
```

Comment:
👍 I think doing the free mem accounting not on every batch iterator was also some issue to come into play or?

Reply:
If I'm not mistaken `apply` is only called once per `ShardingUpsertExecutor` instance.

The main motivation for moving it is to reduce the window between retrieving the free memory from the circuit breaker and accounting for the memory.

I'm actually wondering if the logic now may lead to too many circuit breaker exceptions. If it ends up using the free memory from the circuit breaker because it's lower than the `limit * 0.30`, then with concurrent operations it may be likely that accounting for the free bytes would trip the breaker.


Should we extend the circuitBreaker and have some functionality that allows us to say "I want to reserve X bytes, it that's not available, reserve as much as possible and tell me how much I can use"?

Reply:
I'm not sure if I get you right.
> Should we extend the circuitBreaker and have some functionality that allows us to say "I want to reserve X bytes, if that's not available, reserve as much as possible and tell me how much I can use"?

Isn't this exactly what the code is doing by using the free available mem  (limit - used) if the desired one (limit * 0.3) is not available? 

As the CircuitBreaker's accounting is protected against concurrent issues, how could concurrent operations get wrong free memory and thus result more frequently in breaker exceptions?

Reply:
Getting the free memory via `limit - used` and then calling `addEstimateBytesAndMaybeBreak` is not atomic. In-between the amount of free memory can change. The circuitBreakers concurrency gurantee's are only for the individual calls.

Reply:
I ok now I got it, +1 for moving this into the circuit breaker to make that atomic. I understood this wrong as I interpreted it that the suggestion would result in a different logic.

Reply:
@seut pushed a fixup to make it atomic.